Python/_warnings.c

```diff
@@ -765,6 +899,151 @@ warn_explicit(PyThreadState *tstate, PyObject *category, PyObject *message,
     return result;  /* Py_None or NULL. */
 }
 
+static PyObject *
+warn_explicit_with_fix(PyThreadState *tstate, PyObject *category, PyObject *message,
+              PyObject *fix, PyObject *filename, int lineno,
+              PyObject *module, PyObject *registry, PyObject *sourceline,
+              PyObject *source)
+{
+    PyObject *key = NULL, *text = NULL, *fix_txt = NULL, *result = NULL, *lineno_obj = NULL;
+    PyObject *item = NULL;
+    PyObject *action;
+    int rc;
+    PyInterpreterState *interp = tstate->interp;
+
+    /* module can be None if a warning is emitted late during Python shutdown.
+       In this case, the Python warnings module was probably unloaded, filters
+       are no more available to choose as action. It is safer to ignore the
+       warning and do nothing. */
+    if (module == Py_None)
+        Py_RETURN_NONE;
+
+    if (registry && !PyDict_Check(registry) && (registry != Py_None)) {
+        PyErr_SetString(PyExc_TypeError, "'registry' must be a dict or None");
+        return NULL;
+    }
+
+    /* Normalize module. */
+    if (module == NULL) {
+        module = normalize_module(filename);
+        if (module == NULL)
+            return NULL;
+    }
+    else
+        Py_INCREF(module);
+
+    /* Normalize message. */
+    Py_INCREF(message);  /* DECREF'ed in cleanup. */
+    rc = PyObject_IsInstance(message, PyExc_Warning);
+    if (rc == -1) {
+        goto cleanup;
+    }
+    if (rc == 1) {
+        text = PyObject_Str(message);
+        if (text == NULL)
+            goto cleanup;
+        category = (PyObject*)Py_TYPE(message);
+    }
+    else {
+        text = message;
+        message = PyObject_CallOneArg(category, message);
+        if (message == NULL)
+            goto cleanup;
+    }
+
+    lineno_obj = PyLong_FromLong(lineno);
+    if (lineno_obj == NULL)
+        goto cleanup;
+
+    if (source == Py_None) {
+        source = NULL;
+    }
+
+    /* Create key. */
+    key = PyTuple_Pack(4, text, fix_txt, category, lineno_obj);
```

Comment:
Silly question: if the same error happens multiple times in the same line, this will cause it to be printed only once?

Reply:
Ideally we are supposed to support that, my current tests show otherwise, but the behaviour you describe is the ideal, so I will support that instead.

Reply:
Fixed, this to make sure warning is thrown once